### PR TITLE
Implement ping timeout detection

### PR DESCRIPTION
**Pull request by knitori**
Also fixes some minor issues I had.

Worked fine for me. I tested it by restarting my router :smile_cat:

### CODE REVIEW

File: shanghai/network.py [hunk 1]
```diff
@@ -211,7 +244,7 @@ def sendline(self, line: str):
         self.connection.writeline(line.encode(self.encoding))
 
     def sendcmd(self, command: str, *params: str):
-        args = [command, *params]
+        args = [command, *(str(p) for p in params)]
```
Comment (FichteFoll) [inline shanghai/network.py]: That's kind of a violation of the signature though. I think it is sane to request all parameters to be strings.

File: shanghai/logging.py [hunk 2]
```diff
@@ -24,9 +25,14 @@ def set_logging_config(config):
...
 def _print_like(func):
     @functools.wraps(func)
-    def _wrap(self, *args):
+    def _wrap(self, *args, **kwargs):
         f = io.StringIO()
         print(*args, file=f)
+        if 'exc_info' in kwargs:
+            exc_info = kwargs.pop('exc_info')
+            if exc_info:
+                traceback.print_exc(file=f)
```
Comment (FichteFoll) [inline shanghai/logging.py]: Are you sure this does what you think it does? I mean, `traceback.print_exec` takes the currently being handled exception from `sys` iirc, so passing `exec_info` doesn't mean anything here. In fact, I'm quite sure this is what `.exception` already does, only at the exception level and not at any other (which I don't think we need?).

File: shanghai/network.py (hunk 1 above)
Comment (knitori) [inline shanghai/network.py]: Oh, you're right.

File: shanghai/logging.py (hunk 2 above)
Comment (knitori) [inline shanghai/logging.py]: I don't think anything about that, so I'm pretty sure that I'm unsure about what it does.

I haven't looked further into how `.exception` does things, and forgot about it here.

**Comment by knitori** [comment]
A different method would probably be, to set a property of the time stamp of the last received message, and then somehow check every now and then if the time from the last received message surpasses a set time out. Additionally send a ping packet to test the connection.

Then you'd have to make sure the connection cancelling and sending of the ping packet are not done while offline etc.

File: shanghai/network.py [hunk 3]
```diff
@@ -117,6 +122,31 @@ def start_register(self):
         # TODO add listener for ERR_NICKNAMEINUSE here;
         # maybe also add a listener for RPL_WELCOME to clear this listener
 
+    def ping_timeout(self):
+        current_logger.info('Detected ping timeout.')
+        self.connection_task.cancel()
+
+    def send_ping(self):
+        current_logger.info('Sending ping to test if connection is alive.')
+        self.sendcmd('PING', str(int(time.time())))
```
Comment (FichteFoll) [inline shanghai/network.py]: This will conflict with the other PR, since I renamed the methods to `send_*`.

**Pull request merged**